src/GLSL.hs

```diff
@@ -0,0 +1,75 @@
+module GLSL where
+import Syntax (OpaqueType)
+
+data GSize = Two
+           | Three
+           | Four
+        deriving (Eq, Ord)
+
+data GType = Void
+           | GBool
+           | GInt
+           | GUInt
+           | GFloat
+           | GDouble
+           | GVec GSize GType
+           | GMat GSize GSize
+
+data GOp = GArr
+         | GField
+         | GMult
+         | GMod
+         | GDiv
+         | GAdd
+         | GSub
+         | GRShift
+         | GLShift
+         | GLt
+         | GLte
+         | GGt
+         | GGte
+         | GEq
+         | GNEq
+         | GAnd
+         | GOr
+         | GXOr
+         | GBAnd
+         | GBOr
+         | GBXOr
+
+data GUOp = GFLip
+          | GNot
+          | GNeg
+          | GPos
+
+data GConst = GR String
+            | GI Int
+            | GB Bool
+            | GF Float
+            | GD Double
+            | GV [GConst]
+            | GM [[GConst]]
+
+data GExpr = GCall String [GExpr]
+           | GParen GExpr
+           | GAssign String GExpr
+           | Imm GConst
+           | GUn GUOp GExpr
+           | GBin GOp GExpr GExpr
+           | GTern GExpr GExpr GExpr
+
+data GStmt = GFor GExpr GExpr GExpr [GStmt]
+           | GWhile GExpr [GStmt]
+           | GDoWhile GExpr [GStmt]
+           | GIf [(Maybe GExpr, [GStmt])] -- This is so that we can have if/else if/else chains
+           | GSwitch GExpr [(GConst,[GStmt])]
+           | GStmt GExpr
+           | GRet GExpr
+           | GBreak
+           | GContinue
```

Comment:
Vertext & Fragment shaders can have different kinds of statements. The most noteworthy example is the `discard` statement, which discards a given fragment leaving it 'empty' in a sense. This is used to achieve certain effects, like putting holes in a surface or something similar. To incorporate this, we would at least 3 categories of statements.
```
Stmt -- general statements shared by all shaders
VStmt -- statements only available in the vertex shader (I have to check if there are any...)
FStmt -- statements only available in the fragment shader (at least discard, possibly others)
```